augur/utils.py

```diff
@@ -157,80 +156,11 @@ def read_tree(fname, min_terminals=3):
-                            if n in node_data["nodes"]:
-                                node_data["nodes"][n].update(nv)
-                            else:
-                                node_data["nodes"][n] = nv
-                    elif k=="generated_by":
-                        # Note that this key is _not_ part of the dict returned from this fn.
-                        if v.get("program") == "augur" and not is_augur_version_compatable(v.get("version")):
-                            # check that the augur version, if provided, is compatible.
-                            # ignore version checking of non-augur produced JSONs
-                            raise AugurException("Augur version incompatability detected -- the JSON {} was generated by augur version {} which is "
-                                "incompatable with the current augur version ({}). We suggest you rerun the pipeline using the current "
-                                "version of augur.".format(fname, v.get("version"), get_augur_version()))
-                    elif k in node_data:
-                        # Behavior as of 2019-11-07 is to do a top-level merge
-                        # of dictionaries. If the value is not a dictionary, we
-                        # now have a fatal error with a nice message (note that
-                        # before 2019-11-07 this was an unhandled error).
-                        # This should be revisited in the future. TODO.
-                        if isinstance(node_data[k], dict) and isinstance(v, dict):
-                            node_data[k].update(v)
-                        else:
-                            raise AugurException("\"{}\" key found in multiple JSONs. This is not currently handled by augur, "
-                                "unless all values are dictionaries. "
-                                "Please check the source of these JSONs.".format(k))
-                    else:
-                        node_data[k]=v
-            except AugurException as e:
-                print(e)
-                sys.exit(2)
-        else:
-            print("ERROR: node_data JSON file %s not found. Attempting to proceed without it."%fname)
-
-    if tree and os.path.isfile(tree):
-        try:
-            T = Bio.Phylo.read(tree, 'newick')
-        except:
-            print("Failed to read tree from file "+tree, file=sys.stderr)
-        else:
-            tree_node_names = set([l.name for l in T.find_clades()])
-            meta_node_names = set(node_data["nodes"].keys())
-            if tree_node_names!=meta_node_names:
-                print("Names of nodes (including internal nodes) of tree %s don't"
-                    " match node names in the node data files."%tree, file=sys.stderr)
-    return node_data
+def read_node_data(fnames, tree=None):
+    import augur
+    return augur.util_support.node_data_reader.NodeDataReader(fnames, tree).read()
+    #return NodeDataReader(fnames, tree).read()
```

Comment:
This was necessary to work around a dependency loop:

`augur.util_support.node_data_reader` → `augur.util_support.node_data_file`
`augur.util_support.node_data_file` → `augur.utils`

Importing `augur.util_support.node_data_reader` in this file would create a loop. Using the less specific `import augur` works around it.

In the long run, I would envision `utils.read_node_data` (and `.read_colors`, etc) to go away entirely, and consumers would just use the new class directly. Something like:

```diff
diff --git a/augur/translate.py b/augur/translate.py
index 244b86c1..8da71a67 100644
--- a/augur/translate.py
+++ b/augur/translate.py
@@ -336,7 +336,7 @@ def run(args):
         ref = compress_seq['reference']
         is_vcf = True
     else:
-        node_data = read_node_data(args.ancestral_sequences, args.tree)
+        node_data = NodeDataReader(args.ancestral_sequences, args.tree).read()
         if node_data is None:
             print("ERROR: could not read node data (incl sequences)")
             return 1
```

And then `augur.utils` could become much less of a junk drawer.

Reply:
lol never mind.

The things in `utils` that the new code depends on don't really belong there, anyway. I just moved them and the loop is resolved.